source/extensions/load_balancing_policies/override_host/load_balancer.cc

```diff
@@ -255,49 +269,27 @@ OverrideHostLoadBalancer::LoadBalancerImpl::findHost(absl::string_view endpoint)
-    const std::vector<std::string>& selected_hosts, StreamInfo::FilterState& filter_state) {
-  uint32_t fallback_index = 1;
-  OverrideHostFilterState* override_host_state =
-      filter_state.getDataMutable<OverrideHostFilterState>(
-          OverrideHostFilterState::kFilterStateKey);
-  if (!override_host_state && selected_hosts.size()) {
-    // Use the primary endpoint.
-    ENVOY_LOG(trace, "Selecting primary endpoint {}", selected_hosts[0]);
-    auto new_override_host_state = std::make_shared<OverrideHostFilterState>();
-    filter_state.setData(OverrideHostFilterState::kFilterStateKey, new_override_host_state,
-                         StreamInfo::FilterState::StateType::Mutable);
-    override_host_state = new_override_host_state.get();
-
-    // Endpoint extracted from the header does not have locality.
-    HostConstSharedPtr host = findHost(selected_hosts[0]);
-    // If the primary endpoint was found in the current host set, use it.
-    // Otherwise try to see if one of the failover endpoints is available. This
-    // is possible when the cluster received EDS update while the request to the
-    // endpoint picker was in flight.
-    if (host) {
+    OverrideHostFilterState& override_host_state) {
+  uint64_t host_index = override_host_state.hostIndex();
+  const auto selected_hosts = override_host_state.selectedHosts();
+
+  for (; host_index < selected_hosts.size(); host_index++) {
+    if (HostConstSharedPtr host = findHost(selected_hosts[host_index]); host != nullptr) {
+      ENVOY_LOG(trace, "Selected endpoint {}: {}", host_index, selected_hosts[host_index]);
+      // Update the host index in the filter state to the next host. We will start from that
+      // index in the next call to getEndpoint.
+      override_host_state.setHostIndex(host_index + 1);
       return host;
     }
-  } else {
-    fallback_index = override_host_state->fallbackHostIndex();
   }
 
-  // Loop through fallback hosts until we find one that is available.
-  HostConstSharedPtr host;
-  for (; fallback_index < selected_hosts.size() && !host; ++fallback_index) {
-    ENVOY_LOG(trace, "Selecting failover endpoint {}: {}", fallback_index,
-              selected_hosts[fallback_index]);
-    host = findHost(selected_hosts[fallback_index]);
-  }
+  // If we reach here, it means that all the selected hosts are not available or have used.
+  ASSERT(host_index == selected_hosts.size());
+  override_host_state.setHostIndex(host_index);
+  ENVOY_LOG(trace, "Number of retry attempts has exceeded the number of failover endpoints {}",
+            selected_hosts.size());
 
-  // Update the fallback index in the metadata.
-  override_host_state->setFallbackHostIndex(fallback_index);
-  if (!host) {
-    ENVOY_LOG(trace,
-              "Number of retry attempts {} has exceeded the number of failover "
-              "endpoints {}",
-              fallback_index + 1, selected_hosts.size());
-  }
-  return host;
+  return nullptr;
```

Comment:
We needn't handle the `selected_hosts[0]` and other hosts separately because we will fallback to subseqent hosts anyway if the `selected_hosts[0]` is invalid.
So, simple iteration could be used here.